tee-worker/omni-executor/evm-aggregators-client/src/evm_tx_manager.rs

```diff
@@ -0,0 +1,459 @@
+	SERVICE_FEE_BPS, SERVICE_FEE_PERCENT,
+};
+use crate::inch_client::client::InchSwap;
+use crate::inch_client::types::{convert_slippage_to_inch, SwapRequest};
+use crate::kyber_client::client::KyberSwap;
+use crate::okx_client::client::OkxSwap;
+use ethereum_rpc::RpcProvider;
+use rust_decimal::prelude::{Decimal, ToPrimitive};
+use std::str::FromStr;
+use std::sync::Arc;
+
+use crate::kyber_client::types::GetSwapRouteRequest;
+use crate::okx_client::types::{convert_slippage_to_okx, get_okx_gas_level};
+use alloy::primitives::Uint;
+use alloy::{
+	primitives::{Address, TxKind},
+	rpc::types::{TransactionInput, TransactionRequest},
+};
+use async_trait::async_trait;
+use ethereum_rpc::client::EthereumClient;
+use hex::FromHex;
+use log::error;
+
+/// EVM Transaction Manager
+pub struct EvmTxManager<
+	EthClient: RpcProvider + ?Sized,
+	InchClient: InchSwap + ?Sized,
+	KyberClient: KyberSwap + ?Sized,
+	OkxClient: OkxSwap + ?Sized,
+> {
+	pub okx_client: Arc<OkxClient>,
+	pub kyber_client: Arc<KyberClient>,
+	pub inch_client: Arc<InchClient>,
+	pub eth_client: Arc<EthClient>,
+	pub fee_receiver: String,
+}
+
+#[async_trait]
+pub trait ConstructEvmTx: Send + Sync {
+	async fn construct_inch_tx(
+		&self,
+		create_market_tx: CreateMarketTx,
+		nonce: u64,
+		amount_decimal: Decimal,
+	) -> Result<TransactionRequest, ()>;
+	async fn construct_okx_tx(
+		&self,
+		create_market_tx: CreateMarketTx,
+		nonce: u64,
+		amount_decimal: Decimal,
+	) -> Result<TransactionRequest, ()>;
+	async fn construct_kyber_tx(
+		&self,
+		create_market_tx: CreateMarketTx,
+		nonce: u64,
+		amount_decimal: Decimal,
+	) -> Result<TransactionRequest, ()>;
+}
+
+#[async_trait]
+impl<EthClient, InchClient, KyberClient, OkxClient> ConstructEvmTx
+	for EvmTxManager<EthClient, InchClient, KyberClient, OkxClient>
+where
+	EthClient: RpcProvider + ?Sized,
+	InchClient: InchSwap + ?Sized,
+	KyberClient: KyberSwap + ?Sized,
+	OkxClient: OkxSwap + ?Sized,
+{
+	async fn construct_inch_tx(
+		&self,
+		create_market_tx: CreateMarketTx,
+		nonce: u64,
+		amount_decimal: Decimal,
+	) -> Result<TransactionRequest, ()> {
+		let chain_id = create_market_tx.chain_id;
+
+		let mut swap_request = SwapRequest {
+			chain_id,
+			amount: amount_decimal.to_string(),
+			from_token_address: create_market_tx.in_token_ca.clone(),
+			to_token_address: create_market_tx.out_token_ca.clone(),
+			slippage: convert_slippage_to_inch(create_market_tx.slippage),
+			user_wallet_address: create_market_tx.user_wallet_address.clone(),
+			fee_percent: SERVICE_FEE_PERCENT.to_string(),
+			dex_ids: INCH_DEX_IDS_MAP[&chain_id][&create_market_tx.trade_pool_name].to_string(),
+			referrer: self.fee_receiver.clone(),
+			..Default::default()
+		};
+
+		if is_native_token(&create_market_tx.in_token_ca.into_bytes()) {
+			swap_request.from_token_address = NATIVE_ADDRESS.to_string();
+		} else {
+			swap_request.to_token_address = NATIVE_ADDRESS.to_string();
+		}
```

Comment:
I don't quite understand this logic - then what is `to_token_address: create_market_tx.out_token_ca.clone()` used for?

Reply:
I took this logic as is from the golang implementation, I assume the if the `in_token` is not `NATIVE_TOKEN` then it is assumed that `out_token` is a `NATIVE_TOKEN`, as the comment read they only support to and fro from Native tokens.